plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+        //(1)
+        StringBuffer cmd2("GETSET %b %b");
+        RedisPlugin::appendExpire(cmd2, expire);
+        //obtain channel
+        assertRedisErr(redisAsyncCommand(context, getCB, (void*)&retVal, cmd2.str(), key, strlen(key), value, size), "SET buffer write error");
+        handleLoop(EV_DEFAULT_ EV_TIMEOUT);
+        if (strncmp(retVal.str(), REDIS_LOCK_PREFIX, strlen(REDIS_LOCK_PREFIX)) == 0 )
+        {
+            assertRedisErr(redisAsyncCommand(context, pubCB, NULL, "PUBLISH %b %b", channel, strlen(channel), value, size), "PUBLISH buffer write error");
+            handleLoop(EV_DEFAULT_ EV_TIMEOUT);//again not necessary, could just call redisAsyncHandleWrite(context);
+        }
+    }
+}
+//-------------------------------------------GET-----------------------------------------
+//---OUTER---
+template<class type> void AsyncRGet(ICodeContext * ctx, const char * options, const char * key, type & returnValue, const char * channel, unsigned __int64 database)
+{
+    Owned<AsyncConnection> master = AsyncConnection::createConnection(ctx, options, database);
+    ReturnValue retVal;
+    master->get(ctx, key, &retVal, channel);
+    StringBuffer keyMsg = getFailMsg;
+
+    size_t returnSize = retVal.getSize();
+    if (sizeof(type)!=returnSize)
+    {
+        VStringBuffer msg("RedisPlugin: ERROR - Requested type of different size (%uB) from that stored (%uB).", (unsigned)sizeof(type), (unsigned)returnSize);
+        rtlFail(0, msg.str());
+    }
+    memcpy(&returnValue, retVal.str(), returnSize);
+}
+template<class type> void AsyncRGet(ICodeContext * ctx, const char * options, const char * key, size_t & returnLength, type * & returnValue, const char * channel, unsigned __int64 database)
+{
+    Owned<AsyncConnection> master = AsyncConnection::createConnection(ctx, options, database);
+    master->get(ctx, key, returnLength, returnValue, channel);
+}
+void AsyncRGetVoidPtrLenPair(ICodeContext * ctx, const char * options, const char * key, size_t & returnLength, void * & returnValue, const char * channel, unsigned __int64 database)
+{
+    Owned<AsyncConnection> master = AsyncConnection::createConnection(ctx, options, database);
+    master->getVoidPtrLenPair(ctx, key, returnLength, returnValue, channel);
+}
+//---INNER---
+void AsyncConnection::get(ICodeContext * ctx, const char * key, ReturnValue * retVal, const char * channel )
+{
+    handleLockForGet(ctx, key, channel, retVal);
+}
+template<class type> void AsyncConnection::get(ICodeContext * ctx, const char * key, size_t & returnSize, type * & returnValue, const char * channel)
+{
+    ReturnValue retVal;
+    handleLockForGet(ctx, key, channel, &retVal);
+
+    returnSize = retVal.getSize();
+    returnValue = reinterpret_cast<type*>(retVal.getStr());
+}
+void AsyncConnection::getVoidPtrLenPair(ICodeContext * ctx, const char * key, size_t & returnSize, void * & returnValue, const char * channel )
```

Comment:
I'll look back into it as I can't quite remember. I had compiler/linker issues when using the templates with void*, he wrong function was being called. This was either _me_ having issues or it's an actual issue. I special cased to see what was going on, it worked, I left it.